hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/security/OMRenewDelegationTokenRequest.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.hadoop.ozone.om.response.security.OMRenewDelegationTokenResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.RenewDelegationTokenResponseProto;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.UpdateRenewDelegationTokenRequest;
+import org.apache.hadoop.ozone.protocolPB.OMPBHelper;
+import org.apache.hadoop.ozone.security.OzoneTokenIdentifier;
+import org.apache.hadoop.security.proto.SecurityProtos.RenewDelegationTokenRequestProto;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+/**
+ * Handle RenewDelegationToken Request.
+ */
+public class OMRenewDelegationTokenRequest extends OMClientRequest {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMGetDelegationTokenRequest.class);
+
+  public OMRenewDelegationTokenRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    RenewDelegationTokenRequestProto renewDelegationTokenRequest =
+        getOmRequest().getRenewDelegationTokenRequest();
+
+    // Call OM to renew token
+    long renewTime = ozoneManager.renewDelegationToken(
+        OMPBHelper.convertToDelegationToken(
+            renewDelegationTokenRequest.getToken()));
+
+    RenewDelegationTokenResponseProto.Builder renewResponse =
+        RenewDelegationTokenResponseProto.newBuilder();
+
+    renewResponse.setResponse(org.apache.hadoop.security.proto.SecurityProtos
+        .RenewDelegationTokenResponseProto.newBuilder()
+        .setNewExpiryTime(renewTime));
+
+
+    // Client issues RenewDelegationToken request, when received by OM leader
+    // it will renew the token. Original RenewDelegationToken request is
+    // converted to UpdateRenewDelegationToken request with the token and renew
+    // information. This updated request will be submitted to Ratis. In this
+    // way delegation token renewd by leader, will be replicated across all
+    // OMs. With this approach, original RenewDelegationToken request from
+    // client does not need any proto changes.
+
+    // Create UpdateRenewDelegationTokenRequest with original request and
+    // expiry time.
+    OMRequest.Builder omRequest = OMRequest.newBuilder()
+        .setUserInfo(getUserInfo())
+        .setUpdatedRenewDelegationTokenRequest(
+            UpdateRenewDelegationTokenRequest.newBuilder()
+                .setRenewDelegationTokenRequest(renewDelegationTokenRequest)
+                .setRenewDelegationTokenResponse(renewResponse))
+        .setCmdType(getOmRequest().getCmdType())
```

Comment:
Should we use the cmd type of UpdateDelegationTokenRequest

Reply:
This is being done intentionally so that we don't need new classes or some special handling in OMRatisUtils createOMClientRequest().

Same is explained in the comments.

Reply:
An UpdateRenewDelegationTokenRequest body with RenewDelegationTokenRequest type in OMRequest is a bit hacky. But I'm OK with it if otherwise takes too much changes.

Reply:
Ya can understand that. Otherwise, we need some changes in createClientRequest in ratis to act upon UpdateRenewDelegationTokenRequest or RenewDelegationTokenRequest to return OMGetDelegationTokenRequest object or have separate new classes for both of them. And for one only preExecute will be needed, for other validateAndUpdateCache only will be needed. So, I will leave this as it is.

Reply:
I agree this is a bit hard to follow. I also had a long discussion with Bharat while reviewing the createDelegationToken patch. I am okay to live with for now to avoid creating yet another request class.